### ISSUE TITLE
Scylla repair overstreaming data 

### ISSUE THREAD

**Issue by vabezruchko** [scylladb/scylla#3134]
Installation details
Scylla version (or git commit hash): 2.0.2-0.20171130.07b039f-1
Cluster size:
OS (Debian 8): 32

We noticed that when we run command "nodetool repair -pr" data was overstremed than we expected (up to 50 % of data size).  Actually all data in cluster are consistent (because we ran repair command before) and we expected no data to stream. 

Also we afraid that when data size will grow, daily routine repair on each node will be too long.

Steps for reproduce:

we have 2 tables, schema:
```
CREATE TYPE chats.dialog_type (
    user_id1 bigint,
    user_id2 bigint
);

CREATE TYPE chats.message_attachment (
    entity bigint,
    mime text,
    meta text,
    lat float,
    lon float
);

CREATE TYPE chats.message_body (
    sender bigint,
    type int,
    msg_text text,
    random_id bigint,
    date timestamp,
    ttl int
);

CREATE TABLE chats.message_legacy (
    msg_old_id bigint PRIMARY KEY,
    chat_id bigint,
    date timestamp,
    dialog frozen<dialog_type>,
    msg_id bigint
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy', 'max_threshold': '32', 'min_threshold': '4'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 8640000
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048;

CREATE TABLE chats.dialog_messages_2018 (
    dated int,
    user_id1 bigint,
    user_id2 bigint,
    msg_id bigint,
    attachment frozen<message_attachment>,
    body frozen<message_body>,
    deleted set<bigint>,
    msg_old_id bigint,
    read int,
    received int,
    valid_till timestamp,
    PRIMARY KEY ((dated, user_id1, user_id2), msg_id)
) WITH CLUSTERING ORDER BY (msg_id DESC)
    AND bloom_filter_fp_chance = 0.1
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.DateTieredCompactionStrategy', 'max_threshold': '32', 'min_threshold': '4', 'max_sstable_age_days': '30'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.DeflateCompressor'}
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048;
```
Before run repair we checked table size with following results:
```
6.4G	dialog_messages_2018-6643e510ea2311e7b398000000000003
9.8G	message_legacy-8e7c1160ea2311e7b398000000000003
```
Run command "nodetool repair -pr" at the one node
```
root@xen-scylla-chatmsg01:~# nodetool repair --partitioner-range
[2018-01-19 10:38:27,330] Starting repair command #3, repairing 1 ranges for keyspace chats (parallelism=SEQUENTIAL, full=true)
[2018-01-19 11:23:38,794] Repair session 3
[2018-01-19 11:23:38,795] Repair session 3 finished
[2018-01-19 11:23:38,855] Starting repair command #4, repairing 1 ranges for keyspace system_traces (parallelism=SEQUENTIAL, full=true)
[2018-01-19 11:23:48,971] Repair session 4
[2018-01-19 11:23:48,973] Repair session 4 finished
```
After run repair we check table size with following results:
```
6.5G dialog_messages_2018-6643e510ea2311e7b398000000000003
14G	message_legacy-8e7c1160ea2311e7b398000000000003
```
Data for table message_legacy  was increased nearly 50%. Executed command "nodetool compact chats message_legacy" with following results:
```
9.8G	message_legacy-8e7c1160ea2311e7b398000000000003
```
Also during process repairing table size for dialog_messages was increased up to 9G.

![image](https://user-images.githubusercontent.com/5371224/35144695-563ceee0-fd16-11e7-8282-db45e96c7100.png)

![image](https://user-images.githubusercontent.com/5371224/35144713-68baaa80-fd16-11e7-87fe-71100738d82a.png)

![image](https://user-images.githubusercontent.com/5371224/35144723-6f8b6e26-fd16-11e7-9be4-7bebb218774c.png)

I attached scylla logs.
[scylla_repair_node_log.txt.zip](https://github.com/scylladb/scylla/files/1646020/scylla_repair_node_log.txt.zip)

**Comment by dorlaor** [comment]
Hi Vadim,

We currently have a deficiency in the streaming+repair algorithm and thus we
send way too much data than what it is actually required.
Gleb and Asias are working on an improved algorithm and implementation.
We'll have a solution in release 2.3
They can provide more details.

Dor

**Comment by dorlaor** [comment]
On Tue, Jan 23, 2018 at 7:33 AM, Vadim Bezruchko <notifications@github.com>
wrote:

> Hi Dor Laor,
>
> the Cassandra best-practice manuals (see this
> <http://docs.datastax.com/en/cassandra/2.0/cassandra/operations/ops_repair_nodes_c.html>)
> recommend running a repair at regular intervals.
>

Definitely run frequent repairs

> Now, my experiments show 50% growing data size during the repair
> operation. CPU consumption is very high and operation time is 1 hour just
> for one node. Estimation for total repair is 32 hours for the whole cluster
> because we have 32 nodes. Also, with the increasing amount of data will
> increase the repair time!
>
It's possible to run repair in parallel, we're about to release Scylla
manager which will be the main console to manage
Scylla. It's first main feature is managed, parallel repair. It streams
very small ranges in order to have a controlled effect
on performance (cpu/latency wise). Scylla manager is far more granular than
the nodetool repair.
It's not a must to have 50% increase post-repair, pre-compaction, it's
mainly a matter of fit into the current
compaction policy thresholds. Maybe Raphael/Asias can share more light here.


> Additionally, we cannot abandon the usage of the repair because
> hinted-handoff mechanism has not been implemented yet for scylladb. And for
> every restart node, we need to run the repair.
>
2.2 will have hinted handoff (already committed) but even with it, repair
is always recommended.


> Could you give recommendations to reduce repair overloads or explain how
> it works in productions?
>
- Use either our Scylla Manager or run nodetool repair per keyspace/subrange
- Wait for 2.3 were we'll stream data more efficiently and reduce the
network traffic dramatically.

**Comment by asias** [comment]
Hello @vabezruchko, current scylla repair works at 100 of partitions granularity, i.e., if a single row in that 100 partitions diffs, the whole 100 of partitions will be synced by the repair master by fetching all the rows from participant node and merging them and sending them back. This is why you see disk usage goes up and you see bunch of streaming on wire. The compaction will reduce the disk usage eventually. 

We are working on a new row level repair which will work on the single row granularity which can reduces the unnecessary traffic.

**Comment by slivne** [comment]
@vabezruchko - if the information in the nodes is in synch scylla will not stream data and write it to disk - so if your data was in synch this should not happen.

If there is a workload in parallel that is writing data to the cluster - then there will be "some" difference between the nodes - that is related to the in flight writes.

If your partitions are large and you are appending information to those partitions then as asias explained this will cause sending of all the partition information even if there is only a difference of a single missing row.

You can check partition sizes using nodetool cfstats

With regards to the additional data on the disk - selection of compaction strategy influences the data duplication on disk - its seems you are using STCS and running major compaction.

**Comment by slivne** [comment]
@vabezruchko can you please assist us with answers ...

**Comment by vabezruchko** [comment]
My apologies, unfortunately I missed the question.

We had two problems and we solved those. 

1. for table 'dialog_messages_2018'  we configured DateTieredCompactionStrategy. When we ran the command 'nodetool repair' the node was got very small amount of data from other nodes and initiated unnecessary compaction which resulted unnecessary CPU overload. We solved this problem:  because our table splitted by time periods, we changed compaction strategy to STCS and tuned it's parametres.

2. but for table 'message_legacy' we already configured STCS. We completely don't understand why scylla send a lot of data for this table. You can check schema above (which i attached with first message) and make sure there are no large partitions because there are no clustering key. But we had a lot of partition keys for this table and we had supposed that is our problem. We rewrote our schema and declined to this table.

**Comment by slivne** [comment]
@vabezruchko can you please clarify 

"If there is a workload in parallel that is writing data to the cluster - then there will be "some" difference between the nodes - that is related to the in flight writes."

do you have ongoing writes ?


with regards to "We rewrote our schema and declined to this table." - it seems you have done something to overcome this - can you clarify ?

**Comment by slivne** [comment]
@vabezruchko - ping 

Please note that we are working on #3033 that will provide better performance.

**Comment by slivne** [comment]
closing - no additional information

will be solved with #3033
